### PR TITLE
feature(#626): Added feature pipeline and horizon splitting

### PR DESCRIPTION
**Pull request by egordm**
This pull request introduces a feature engineering pipeline for multi-horizon time series forecasting models and adds support for horizon-aware transformations across the codebase. The main changes include the implementation of the `FeaturePipeline` class, enhancements to the transform interface to support multi-horizon fitting and transformation, and updates to several general transforms to leverage the new horizon-aware interface.

**Feature engineering pipeline and multi-horizon support:**

* Added the `FeaturePipeline` class to orchestrate versioned and horizon-specific feature transformations for multi-horizon forecasting workflows. The pipeline manages fitting and transforming both versioned and unversioned datasets and ensures correct handling of horizon-specific logic. [[1]](diffhunk://#diff-0bddee4ebc6862175effa54222ab7693dec46afa0b7a4ec30355dfa64dc84132R1-R172) [[2]](diffhunk://#diff-fe7c7456f13b6d4b24b11880ced228a565bc4cfab01493ed11be89faf9fadb55R11-R32)
* Introduced the `fit_horizons`, `transform_horizons`, and `fit_transform_horizons` methods to the `TimeSeriesTransform` interface, enabling transforms to process data across multiple forecast horizons in a unified way.
* Added the `MultiHorizonTimeSeriesDataset` type alias for clarity when working with horizon-indexed datasets.

**General transforms:**

* Updated `ClippingTransform`, `ImputationTransform`, `RemoveEmptyColumnsTransform`, and `ScalerTransform` to implement the new `fit_horizons` method. These transforms now concatenate horizon datasets row-wise when fitting, ensuring global statistics are computed across all horizons. [[1]](diffhunk://#diff-20fad7934baa869405965d9103bd19d8a6412d7ddb27c1cb9adeed28a15d0694R70-R74) [[2]](diffhunk://#diff-e90ceb245c27a7615a281e05187342bc5a29aa1b22fa385635679d11d6552c39R148-R153) [[3]](diffhunk://#diff-cbfa6dd1ac0577057074e927b6b3423a4c3b3a92c938f3ddd1f0f7978f8060c8R84-R88) [[4]](diffhunk://#diff-4161568fd836d2e2b2e01a4b08f6add02edb7d46c8507be10931ce59e101916eR87-R91)
* Added necessary imports for `LeadTime` and the `concat_horizon_datasets_rowwise` utility in the corresponding transform modules. [[1]](diffhunk://#diff-20fad7934baa869405965d9103bd19d8a6412d7ddb27c1cb9adeed28a15d0694R21-R22) [[2]](diffhunk://#diff-e90ceb245c27a7615a281e05187342bc5a29aa1b22fa385635679d11d6552c39R20-R23) [[3]](diffhunk://#diff-cbfa6dd1ac0577057074e927b6b3423a4c3b3a92c938f3ddd1f0f7978f8060c8R21-R22) [[4]](diffhunk://#diff-4161568fd836d2e2b2e01a4b08f6add02edb7d46c8507be10931ce59e101916eR21-R22)

**Core transform infrastructure:**

* Renamed `TimeSeriesVersionedTransform` to `VersionedTimeSeriesTransform` throughout the codebase for clarity and consistency. [[1]](diffhunk://#diff-2632edfa6ce67dd845cf19c0ca6ed783c45842b213e8953b5a767633953d65cdL100-R152) [[2]](diffhunk://#diff-2632edfa6ce67dd845cf19c0ca6ed783c45842b213e8953b5a767633953d65cdL163-R215)

**Other:**

* Removed the unused copyright and docstring header from `exceptions.py`.

### CODE REVIEW

File: packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        ...     horizon_transforms=[
+        ...         CyclicFeaturesTransform(included_features=["timeOfDay", "season"])
+        ...     ]
+        ... )
+        >>>
+        >>> # Process the versioned dataset through the pipeline
+        >>> horizon_datasets = pipeline.fit_transform(versioned_dataset)
+        >>> len(horizon_datasets)
+        2
+        >>> list(horizon_datasets.keys())  # doctest: +SKIP
```
Comment (MvLieshout) [inline packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py]: Why do we skip this one? Does it take a long time?

File: packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        >>> single_pipeline = FeaturePipeline(
+        ...     horizons=[LeadTime.from_string("PT36H")],
+        ...     horizon_transforms=[
+        ...         CyclicFeaturesTransform(included_features=["timeOfDay"])
+        ...     ]
+        ... )
+        >>>
+        >>> # Process the simple dataset through the pipeline
+        >>> single_result = single_pipeline.fit_transform(simple_dataset)
+        >>> len(single_result)
```
Comment (MvLieshout) [inline packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py:113]: Still think it would be nice to show actual dataframes here (could also be another ticket).

File: packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    [
+        pytest.param([LeadTime.from_string("PT1H")], [], [], {"load", "temperature"}, id="no_transforms"),
+        pytest.param(
+            [LeadTime.from_string("PT1H")],
+            [MockVersionedTimeSeriesTransform("_v")],
+            [MockTimeSeriesTransform("_h")],
+            {"load_v_h", "temperature_v_h"},
+            id="all_transforms",
+        ),
+        pytest.param(
```
Comment (MvLieshout) [inline packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py:126]: Nice test. I think we could also add the case where there are multiple horizons and also versioned and unversioned transforms.

File: packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    result = pipeline.transform(sample_timeseries_dataset)
+
+    # Assert
+    assert pipeline._is_fitted
+    assert len(result) == 1
+    horizon_dataset = result[LeadTime.from_string("PT1H")]
+    assert set(horizon_dataset.data.columns) == {"load_test", "temperature_test"}
+
+
+def test_feature_pipeline_unversioned_dataset_validation_errors(sample_timeseries_dataset: TimeSeriesDataset):
```
Comment (MvLieshout) [inline packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py]: I feel like this is actually two tests in one. Might be better to split it (also for maintainability).

File: packages/openstef-models/tests/unit/feature_engineering/test_horizon_split_transform.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            "2025-01-01T12:05:00",
+            "2025-01-01T13:05:00",
+        ]),
+        "load": [100.0, 110.0, 120.0, 130.0],
+        "temperature": [20.0, 21.0, 22.0, 23.0],
+    })
+    return VersionedTimeSeriesDataset.from_dataframe(data, timedelta(hours=1))
+
+
+@pytest.mark.parametrize(
```
Comment (MvLieshout) [inline packages/openstef-models/tests/unit/feature_engineering/test_horizon_split_transform.py:61]: To me it was somewhat confusing that these tests generate dataframes that consist only of NaNs. Is that on purpose?

File: packages/openstef-models/tests/unit/feature_engineering/validation_transforms/test_flatliner_check.py [hunk 6]
```diff
@@ -115,14 +109,14 @@ def test_different_load_column_name() -> None:
...
-    assert transform.is_flatliner_detected is True
+    # Act & Assert
+    try:
+        transform.transform(data)
+        pytest.fail("FlatlinerDetectedError was not raised as expected.")
+    except FlatlinerDetectedError:
+        ...
 
 
 def test_fit_raises_on_missing_load_column() -> None:
```
Comment (MvLieshout) [inline packages/openstef-models/tests/unit/feature_engineering/validation_transforms/test_flatliner_check.py]:
```suggestion
def test_transform_raises_on_missing_load_column() -> None:
```

File: packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py (hunk 1 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py]: Fixed. Needed a good repr for LeadTime

File: packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py (hunk 2 above)
Comment (egordm) [reply packages/openstef-models/src/openstef_models/feature_engineering/feature_pipeline.py:113]: Also fixed. I print now the column names. Actual values are irrelevant, since that's more transform centric.

But it's now showing which horizons it creates data for, and then what the columns are.

File: packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py (hunk 3 above)
Comment (egordm) [reply packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py:126]: Fixed.

File: packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py (hunk 4 above)
Comment (egordm) [reply packages/openstef-models/tests/unit/feature_engineering/test_feature_pipeline.py]: Fixed.

File: packages/openstef-models/tests/unit/feature_engineering/test_horizon_split_transform.py (hunk 5 above)
Comment (egordm) [reply packages/openstef-models/tests/unit/feature_engineering/test_horizon_split_transform.py:61]: Fixed. Using predefined horizon data with labels now.
